src/mcpwm.rs

```diff
@@ -0,0 +1,527 @@
+//! let mcpwm = Mcpwm::new(peripherals.mcpwm0.mcpwm)?;
+//! let mut operator = Operator::new(
+//!     peripherals.mcpwm0.operator0,
+//!     &mcpwm,
+//!     &config,
+//!     peripherals.pins.gpio4,
+//!     peripherals.pins.gpio5,
+//! )?;
+//!
+//! println!("Starting duty-cycle loop");
+//!
+//! for &duty in [0.0, 20.0, 40.0, 60.0, 80.0, 100.0].iter() {
+//!     println!("Duty {}%", duty);
+//!     operator.set_duty_a(duty)?;
+//!     operator.set_duty_b(100.0 - duty)?;
+//!     FreeRtos.delay_ms(2000)?;
+//! }
+//! ```
+//!
+//! See the `examples/` folder of this repository for more.
+
+use core::borrow::Borrow;
+
+use crate::gpio::OutputPin;
+use crate::units::{FromValueType, Hertz};
+use esp_idf_sys::{esp, mcpwm_config_t, EspError};
+use esp_idf_sys::{
+    mcpwm_counter_type_t,
+    mcpwm_counter_type_t_MCPWM_DOWN_COUNTER,
+    //mcpwm_counter_type_t_MCPWM_FREEZE_COUNTER,
+    mcpwm_counter_type_t_MCPWM_UP_COUNTER,
+    mcpwm_counter_type_t_MCPWM_UP_DOWN_COUNTER,
+};
+use esp_idf_sys::{
+    mcpwm_duty_type_t, mcpwm_duty_type_t_MCPWM_DUTY_MODE_0, mcpwm_duty_type_t_MCPWM_DUTY_MODE_1,
+};
+use esp_idf_sys::{
+    mcpwm_io_signals_t, mcpwm_io_signals_t_MCPWM0A, mcpwm_io_signals_t_MCPWM0B,
+    mcpwm_io_signals_t_MCPWM1A, mcpwm_io_signals_t_MCPWM1B, mcpwm_io_signals_t_MCPWM2A,
+    mcpwm_io_signals_t_MCPWM2B,
+};
+use esp_idf_sys::{
+    mcpwm_timer_t, mcpwm_timer_t_MCPWM_TIMER_0, mcpwm_timer_t_MCPWM_TIMER_1,
+    mcpwm_timer_t_MCPWM_TIMER_2,
+};
+use esp_idf_sys::{mcpwm_unit_t, mcpwm_unit_t_MCPWM_UNIT_0, mcpwm_unit_t_MCPWM_UNIT_1};
+
+/// The Motor Control Pulse Width Modulator peripheral
+pub struct Peripheral<U: Unit> {
+    pub mcpwm: MCPWM<U>, // TODO: Is there a better way to name or structure this?
+    pub operator0: OPERATOR0<U>,
+    pub operator1: OPERATOR1<U>,
+    pub operator2: OPERATOR2<U>,
+}
+
+impl<U: Unit> Peripheral<U> {
+    /// # Safety
+    ///
+    /// It is safe to instantiate this exactly one time per `Unit`.
+    pub unsafe fn new() -> Self {
+        Self {
+            mcpwm: MCPWM::new(),
+            operator0: OPERATOR0::new(),
+            operator1: OPERATOR1::new(),
+            operator2: OPERATOR2::new(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum DutyMode {
+    ActiveHigh,
+    ActiveLow,
+}
+
+impl From<DutyMode> for mcpwm_duty_type_t {
+    fn from(val: DutyMode) -> Self {
+        match val {
+            DutyMode::ActiveHigh => mcpwm_duty_type_t_MCPWM_DUTY_MODE_0,
+            DutyMode::ActiveLow => mcpwm_duty_type_t_MCPWM_DUTY_MODE_1,
+        }
+    }
+}
+
+// TODO: For UpDown, frequency is half of MCPWM frequency set
+#[derive(Clone, Copy, Debug)]
+pub enum CounterMode {
+    //Frozen,
+    Up,
+    Down,
+    UpDown,
+}
+
+impl From<CounterMode> for mcpwm_counter_type_t {
+    fn from(val: CounterMode) -> Self {
+        match val {
+            // TODO: This seems to be new to IDF 4.4?
+            //CounterMode::Frozen => mcpwm_counter_type_t_MCPWM_FREEZE_COUNTER,
```

Comment:
We can `cfg` this with `#[cfg(idf_newer_than_or_equal_v4_4_0)]`

Reply:
There is `#[cfg(esp_idf_version = "4.3")]` or `#[cfg(esp_idf_version_major = "4")]`.

>  I believe ESP32-S3 requires IDF >= v4.4

This could be correct, 4.3 only had partial support for S3 so 4.3 may not have the MCPWM implemented then.

Reply:
Oh nice, I will give that a try :)

Reply:
What IDF versions are supported. What versions do I need to check? Would simply `#[cfg(not(esp_idf_version = "4.3"))]` be enough or do we care about older versions?

Reply:
We have been defaulting to v4.4 for well over 6months now, and 4.3 is no longer getting Rust backports so let's focus on 4.4

Reply:
> We have been defaulting to v4.4 for well over 6months now [...]

On second thought, during #96 I had to use `export ESP_IDF_VERSION=release/v4.4;` before every cargo command or it would try to build it as `v4.3.2`. Is this how it should be?